File: examples/go/04-search-text/main.go
// Search for a term across all pages of a PDF and print matches.
// Run: go run main.go document.pdf "query"

package main

import (
	"fmt"
	"log"
	"os"

	"github.com/yfedoseev/pdf_oxide/go"
)

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintln(os.Stderr, "Usage: go run main.go <file.pdf> <query>")
		os.Exit(1)
	}

	path := os.Args[1]
	query := os.Args[2]

	doc, err := pdfoxide.Open(path)
	if err != nil {
		log.Fatalf("Failed to open %s: %v", path, err)
	}
	defer doc.Close()

	pages, _ := doc.PageCount()
	fmt.Printf("Searching for %q in %s (%d pages)...\n\n", query, path, pages)

	// Build the per-page search index for every page up front, instead of
	// paying for it lazily on the first SearchPage()/SearchAll() call.
	// Worth it here since we're about to search every page anyway.
	if err := doc.PrepareSearch(); err != nil {
		log.Fatalf("PrepareSearch: %v", err)
	}

	total := 0
	pagesWithHits := 0

	for i := 0; i < pages; i++ {
		results, err := doc.SearchPage(i, query, false)
		if err != nil || len(results) == 0 {
			continue
		}
		pagesWithHits++
		fmt.Printf("Page %d: %d match(es)\n", i+1, len(results))
		for _, r := range results {
			fmt.Printf("  - %q (x=%.1f y=%.1f)\n", r.Text, r.X, r.Y)
			total++
		}
		fmt.Println()
	}

	fmt.Printf("Found %d total matches across %d pages.\n", total, pagesWithHits)

	// Free the cached search index now that we're done searching — useful
	// before heavy extraction work on the same document object.
	if err := doc.ClearSearchIndex(); err != nil {
		log.Fatalf("ClearSearchIndex: %v", err)
	}
}

File: examples/python/04-search-text/main.py
# Search for a term across all pages of a PDF and print matches.
# Run: python main.py document.pdf "query"

import sys

from pdf_oxide import PdfDocument


def main():
    if len(sys.argv) < 3:
        print("Usage: python main.py <file.pdf> <query>", file=sys.stderr)
        sys.exit(1)

    path = sys.argv[1]
    query = sys.argv[2]
    doc = PdfDocument(path)

    pages = doc.page_count()
    print(f'Searching for "{query}" in {path} ({pages} pages)...\n')

    # Build the per-page search index for every page up front, instead of
    # paying for it lazily on the first search_page()/search() call. Worth
    # it here since we're about to search every page anyway.
    doc.prepare_search()

    total = 0
    pages_with_hits = 0

    for i in range(pages):
        results = doc.search_page(i, query)
        if not results:
            continue
        pages_with_hits += 1
        print(f"Page {i + 1}: {len(results)} match(es)")
        for r in results:
            print(f'  - "{r["text"]}" (x={r["x"]:.1f} y={r["y"]:.1f})')
            total += 1
        print()

    print(f"Found {total} total matches across {pages_with_hits} pages.")

    # Free the cached search index now that we're done searching — useful
    # before heavy extraction work on the same document object.
    doc.clear_search_index()


if __name__ == "__main__":
    main()
